tasks/actorstate/miner/sector_events.go

```diff
@@ -161,117 +246,136 @@ func ExtractMinerPreCommitEvents(extState extraction.State, preCommits *miner.Pr
-			StateRoot: extState.CurrentTipSet().ParentState().String(),
-			SectorID:  u,
-			Event:     event,
-		})
 		return nil
-	}); err != nil {
-		return nil, fmt.Errorf("walking miners removed sectors: %w", err)
+	})
+	grp.Go(func() error {
+		current, err = LoadSectorState(extState.CurrentState())
+		if err != nil {
+			return fmt.Errorf("loading current sector states %w", err)
+		}
+		return nil
+	})
+	// if either load operation fails abort
+	if err := grp.Wait(); err != nil {
+		return nil, err
 	}
 
-	// track recovering sectors
-	if err := recovering.ForEach(func(u uint64) error {
-		out = append(out, &minermodel.MinerSectorEvent{
-			Height:    int64(extState.CurrentTipSet().Height()),
-			MinerID:   extState.Address().String(),
-			StateRoot: extState.CurrentTipSet().ParentState().String(),
-			SectorID:  u,
-			Event:     minermodel.SectorRecovering,
-		})
-		return nil
-	}); err != nil {
-		return nil, fmt.Errorf("walking miners recovering sectors: %w", err)
+	// previous live sector minus current live sectors are sectors removed this epoch.
+	removed, err := bitfield.SubtractBitField(previous.Live, current.Live)
+	if err != nil {
+		return nil, fmt.Errorf("comparing previous live sectors to current live sectors %w", err)
 	}
 
-	// track faulted sectors
-	if err := faulted.ForEach(func(u uint64) error {
-		out = append(out, &minermodel.MinerSectorEvent{
-			Height:    int64(extState.CurrentTipSet().Height()),
-			MinerID:   extState.Address().String(),
-			StateRoot: extState.CurrentTipSet().ParentState().String(),
-			SectorID:  u,
-			Event:     minermodel.SectorFaulted,
-		})
-		return nil
-	}); err != nil {
-		return nil, fmt.Errorf("walking miners faulted sectors: %w", err)
+	// current recovering sectors minus previous recovering sectors are sectors recovering this epoch.
+	recovering, err := bitfield.SubtractBitField(current.Recovering, previous.Recovering)
+	if err != nil {
+		return nil, fmt.Errorf("comparing current recovering sectors to previous recovering sectors %w", err)
 	}
 
-	// track recovered sectors
-	if err := recovered.ForEach(func(u uint64) error {
-		out = append(out, &minermodel.MinerSectorEvent{
-			Height:    int64(extState.CurrentTipSet().Height()),
-			MinerID:   extState.Address().String(),
-			StateRoot: extState.CurrentTipSet().ParentState().String(),
-			SectorID:  u,
-			Event:     minermodel.SectorRecovered,
-		})
-		return nil
-	}); err != nil {
-		return nil, fmt.Errorf("walking miners recovered sectors: %w", err)
+	// current faulty sectors minus previous faulty sectors are sectors faulted this epoch.
+	faulted, err := bitfield.SubtractBitField(current.Faulty, previous.Faulty)
+	if err != nil {
+		return nil, fmt.Errorf("comparing current faulty sectors to previous faulty sectors %w", err)
 	}
-	return out, nil
+
+	// previous faulty sectors minus current active sectors are sectors recovered this epoch.
+	recovered, err := bitfield.IntersectBitField(previous.Faulty, current.Active)
```

Comment:
The comment and code don't match. I assume we only want to see which previous faults overlap with current actives... this makes sense as being "Recovered". Let's update this comment.

`// previous faulty sectors which match (intersect) active sectors are sectors recovered this epoch.`

Reply:
Whoops, bad copy paste